src/terminal/input/terminalInput.cpp

```diff
@@ -440,120 +440,138 @@ static bool _translateDefaultMapping(const KeyEvent& keyEvent,
     return match.has_value();
 }
 
-bool TerminalInput::HandleKey(const IInputEvent* const pInEvent) const
+// Routine Description:
+// - Sends the given input event to the shell.
+// - The caller should attempt to fill the char data in pInEvent if possible.
+//   The char data should already be translated in accordance to Ctrl/Alt/Shift
+//   modifiers, like the characters given by the WM_CHAR event.
+// - The caller doesn't need to fill in any char data for:
+//   - Tab key
+//   - Alt+key combinations
+// - This method will alias Ctrl+Space as a synonym for Ctrl+@ - the null byte.
+// Arguments:
+// - keyEvent - Key event to translate
+// Return Value:
+// - True if the event was handled.
+bool TerminalInput::HandleKey(const IInputEvent* const pInEvent)

```

Comment:
I'm sorry. 🙈
Most changes to this function where necessary to ensure everything still works when you funnel character events into `HandleKey()` instead of `HandleChar()` (which doesn't exist anymore).

At some point I got so confused by the code that I just decided to rewrite this function _from scratch_.
This also means though that it actually might be best if you review changes to this function by pulling up [the current code](https://github.com/microsoft/terminal/blob/1ae4252a7bc07eb08511c837e572ae0ffa58694c/src/terminal/input/terminalInput.cpp#L443-L548) and comparing it side-by-side with the one in my PR, without viewing the diff.
While the structure is entirely different, you should quickly notice that the logic flow itself is the same as before.
I added comments to each section explaining how I understood the code and why it's necessary.